README.md
# Cryptomute ![test](https://github.com/Teakowa/cryptomute/workflows/test/badge.svg)

A small PHP class implementing Format Preserving Encryption via Feistel Network.

[![Latest Version on Packagist](https://img.shields.io/packagist/v/teakowa/cryptomute.svg?style=flat-square)](https://packagist.org/packages/teakowa/cryptomute)
[![PHP from Packagist](https://img.shields.io/packagist/php-v/teakowa/cryptomute?style=flat-square)](https://packagist.org/packages/teakowa/cryptomute)
[![LICENSE](https://img.shields.io/badge/License-Apache--2.0-green.svg?style=flat-square)](LICENSE)
[![LICENSE](https://img.shields.io/badge/License-Anti%20996-blue.svg?style=flat-square)](https://github.com/996icu/996.ICU/blob/master/LICENSE)
[![996.icu](https://img.shields.io/badge/Link-996.icu-red.svg?style=flat-square)](https://996.icu)
[![Total Downloads](https://img.shields.io/packagist/dt/teakowa/cryptomute.svg?style=flat-square)](https://packagist.org/packages/teakowa/cryptomute)

## 1. Installation

You can install Cryptomute via [Composer](http://getcomposer.org) (packagist has [teakowa/cryptomute](https://packagist.org/packages/teakowa/cryptomute) package). In your `composer.json` file use:

``` json
{
    "require": {
        "teakowa/cryptomute": "^1.4"
    }
}
```

And run: `php composer.phar install`. After that you can require the autoloader and use Cryptomute:

## 2. Usage

``` php
use Cryptomute\Cryptomute;

$cryptomute = new Cryptomute(
    'aes-256-cbc',      // cipher
    '0123456789zxcvbn', // base key
    7,                  // number of rounds
);

$password = '0123456789qwerty';
$iv = '0123456789abcdef';

$plainValue = '2048';
$encoded = $cryptomute->encrypt($plainValue, 10, false, $password, $iv);
$decoded = $cryptomute->decrypt($encoded, 10, false, $password, $iv);

var_dump([
  'plainValue' => $plainValue,
  'encoded'    => $encoded,
  'decoded'    => $decoded,
]);
```

```
array(3) {              
  ["plainValue"]=>       
  string(4) "2048"       
  ["encoded"]=>          
  string(9) "309034283"  
  ["decoded"]=>          
  string(4) "2048"       
}                        
```
	
## 3. Options

### 3.1 Cipher
 
Cipher is the first constructor argument. Supported cipher methods are:

Cipher             | IV
------------------ | ---
`des-cbc`          | yes
`aes-128-cbc`      | yes
`aes-192-cbc`      | yes
`aes-256-cbc`      | yes
`camellia-128-cbc` | yes
`camellia-192-cbc` | yes
`camellia-256-cbc` | yes

### 3.2 Key

Key is the second constructor argument. Base key from which all round keys are derrived.

### 3.3 Rounds

Rounds is the third constructor argument. Must be an odd integer greater or equal to 3. More rounds is more secure,
but also slower. Recommended value is at least 7.

## 4. Public methods

### 4.1 setValueRange(`$minValue`, `$maxValue`)

Sets minimum and maximum values. If the result is out of range it will be re-encrypted (or re-decrypted) until output
is in range.

### 4.2 encrypt(`$plainValue`, `$base`, `$pad`, `$password`, `$iv`)
 
Encrypts data. Takes following arguments:

* `$plainValue` (string) input data to be encrypted
* `$base` (int) input data base, accepted values is 2 (binary), 10 (decimal) or 16 (hexadecimal)
* `$pad` (bool) pad left output to match `$maxValue`'s length?
* `$password` (string) encryption password
* `$iv` (string) initialization vector - only if cipher requires it


### 4.2 decrypt(`$cryptValue`, `$base`, `$pad`, `$password`, `$iv`)
 
Decrypts data. Takes following arguments:

* `$cryptValue` (string) input data to be decrypted
* `$base` (int) input data base, accepted values is 2 (binary), 10 (decimal) or 16 (hexadecimal)
* `$pad` (bool) pad left output to match `$maxValue`'s length?
* `$password` (string) encryption password
* `$iv` (string) initialization vector - only if cipher requires it

## LICENSE

The code in this repository, unless otherwise noted, is under the terms of both the [Anti 996](https://github.com/996icu/996.ICU/blob/master/LICENSE) License and the [Apache License (Version 2.0)]().


CHANGELOG.md
<a name="unreleased"></a>
## [Unreleased]


<a name="v1.4.1"></a>
## [v1.4.1] - 2020-08-29
### Code Refactoring
- **DEFAULT_MAX_VALUE:** support 20 digits from 15 digits


<a name="v1.4.0"></a>
## [v1.4.0] - 2020-08-29
### Code Refactoring
- **_hashPassword:** replace hash cipher md5 to sha3-512
- **test:** replace aes 128 to 256


<a name="v1.3.0"></a>
## [v1.3.0] - 2020-08-29
### Features
- **ciphers:** add camellia-256-cbc support


<a name="v1.2.0"></a>
## [v1.2.0] - 2020-08-29
### Code Refactoring
- **ciphers:** remove all ecb ciphers support


<a name="v1.1.0"></a>
## [v1.1.0] - 2020-08-29
### Code Refactoring
- **Cryptomute:** remove const VERSION
- **DEFAULT_MAX_VALUE:** set to 999999999999999 from 9999999999
- **test:** replace expectedException
- **test:** replace openssl_random_pseudo_bytes to random_bytes
- **tests:** replace travis to github action
- **tests:** replace PHPUnit_Framework_TestCase namespace

### Features
- add aes-256-cbc cipher
- add type required


<a name="v1.0.3"></a>
## [v1.0.3] - 2016-07-15

<a name="v1.0.2"></a>
## [v1.0.2] - 2016-05-04
### Pull Requests
- Merge pull request [#14](https://github.com/teakowa/cryptomute/issues/14) from loostro/bugfix-value-range-less-than-input


<a name="v1.0.1"></a>
## [v1.0.1] - 2016-03-03
### Pull Requests
- Merge pull request [#12](https://github.com/teakowa/cryptomute/issues/12) from loostro/bugfix-error-in-comparison-11


<a name="v1.0.0"></a>
## v1.0.0 - 2016-02-17
### Pull Requests
- Merge pull request [#9](https://github.com/teakowa/cryptomute/issues/9) from loostro/feature-more-cipher-methods
- Merge pull request [#7](https://github.com/teakowa/cryptomute/issues/7) from loostro/feature-better-api-and-wider-test-coverage
- Merge pull request [#6](https://github.com/teakowa/cryptomute/issues/6) from loostro/feature-better-api-and-wider-test-coverage
- Merge pull request [#2](https://github.com/teakowa/cryptomute/issues/2) from loostro/feature-add-min-value
- Merge pull request [#1](https://github.com/teakowa/cryptomute/issues/1) from loostro/feature-add-key-and-password-setters


[Unreleased]: https://github.com/teakowa/cryptomute/compare/v1.4.1...HEAD
[v1.4.1]: https://github.com/teakowa/cryptomute/compare/v1.4.0...v1.4.1
[v1.4.0]: https://github.com/teakowa/cryptomute/compare/v1.3.0...v1.4.0
[v1.3.0]: https://github.com/teakowa/cryptomute/compare/v1.2.0...v1.3.0
[v1.2.0]: https://github.com/teakowa/cryptomute/compare/v1.1.0...v1.2.0
[v1.1.0]: https://github.com/teakowa/cryptomute/compare/v1.0.3...v1.1.0
[v1.0.3]: https://github.com/teakowa/cryptomute/compare/v1.0.2...v1.0.3
[v1.0.2]: https://github.com/teakowa/cryptomute/compare/v1.0.1...v1.0.2
[v1.0.1]: https://github.com/teakowa/cryptomute/compare/v1.0.0...v1.0.1


tests/CryptomuteTest.php
<?php

namespace Cryptomute\Tests;

use Cryptomute\Cryptomute;
use InvalidArgumentException;
use PHPUnit\Framework\TestCase;

class CryptomuteTest extends TestCase
{
    public const MIN_VALUE = '0';
    public const MAX_VALUE = '9999999999';
    public const TEST_REPEATS = 15;
    public static $testedCiphers = [
        'des-cbc' => true,
        'aes-128-cbc' => true,
        'aes-192-cbc' => true,
        'aes-256-cbc' => true,
        'camellia-128-cbc' => true,
        'camellia-192-cbc' => true,
        'camellia-256-cbc' => true,
    ];
    public static $testedRounds = [
        3 => 'minimum',
        5 => 'normal',
        7 => 'recommended',
        11 => 'insane',
    ];

    /**
     * @throws \Exception
     */
    public function testReadmeExample()
    {
        $message = '[%s / %d rounds] Expected README example to work.';

        foreach (self::$testedCiphers as $cipher => $usesIv) {
            foreach (self::$testedRounds as $rounds => $description) {
                $cryptomute = $this->getCryptomute($cipher, $rounds);
                $password = random_bytes(32);
                $iv = $usesIv ? random_bytes(openssl_cipher_iv_length($cipher)) : null;

                $plainValue = '2048';
                $encoded = $cryptomute->encrypt($plainValue, 10, false, $password, $iv);
                $decoded = $cryptomute->decrypt($encoded, 10, false, $password, $iv);

                $this->assertEquals($plainValue, $decoded, sprintf(
                    $message,
                    $cipher,
                    $rounds
                ));
            }
        }
    }

    /**
     * @throws \Exception
     */
    public function testEncodedValuesAreInDomain()
    {
        $message = '[%s / %d rounds] Base %d value of %s encoded to %s (%s) is still in domain %s - %s.';
        $minVal = 100;
        $maxVal = 999;

        foreach (self::$testedCiphers as $cipher => $usesIv) {
            foreach (self::$testedRounds as $rounds => $description) {
                $cryptomute = $this->getCryptomute($cipher, $rounds, "$minVal", "$maxVal");
                $password = random_bytes(32);

                foreach (Cryptomute::$allowedBases as $base => $pattern) {
                    for ($i = $minVal; $i <= $maxVal; $i++) {
                        $iv = $usesIv ? random_bytes(openssl_cipher_iv_length($cipher)) : null;

                        $input = gmp_strval(gmp_init("$i", 10), $base);
                        $encoded = $cryptomute->encrypt($input, $base, false, $password, $iv);
                        $intVal = (int) gmp_strval(gmp_init($encoded, $base), 10);

                        $this->assertTrue(
                            $minVal <= $intVal && $intVal <= $maxVal,
                            sprintf($message, $cipher, $rounds, $base, $i, $encoded, "$intVal", $minVal, $maxVal)
                        );
                    }
                }
            }
        }
    }

    /**
     * @throws \Exception
     */
    public function testDecodesEncodedNumbers()
    {
        $message = '[%s / %d rounds] Encoded base %d value must decode to initial value [%s].';

        foreach (self::$testedCiphers as $cipher => $usesIv) {
            foreach (self::$testedRounds as $rounds => $description) {
                $cryptomute = $this->getCryptomute($cipher, $rounds);
                $password = random_bytes(32);

                foreach (Cryptomute::$allowedBases as $base => $pattern) {
                    for ($i = 0; $i < self::TEST_REPEATS; $i++) {
                        $iv = $usesIv ? random_bytes(openssl_cipher_iv_length($cipher)) : null;

                        $input = gmp_strval(gmp_random_range(
                            gmp_init(self::MIN_VALUE, 10),
                            gmp_init(self::MAX_VALUE, 10)
                        ), $base);

                        $encrypted = $cryptomute->encrypt($input, $base, false, $password, $iv);
                        $decrypted = $cryptomute->decrypt($encrypted, $base, false, $password, $iv);

                        $this->assertEquals($input, $decrypted, sprintf(
                            $message,
                            $cipher,
                            $rounds,
                            $base,
                            $input
                        ));
                    }
                }
            }
        }
    }

    /**
     * @throws \Exception
     */
    public function testEncodeWithDiffrentPasswordsProducesDiffrentResults()
    {
        $message = '[%s / %d rounds] Same base %d value encoded with diffrent passwords produce diffrent results [%s].';

        foreach (self::$testedCiphers as $cipher => $usesIv) {
            foreach (self::$testedRounds as $rounds => $description) {
                $cryptomute = $this->getCryptomute($cipher, $rounds);

                foreach (Cryptomute::$allowedBases as $base => $pattern) {
                    for ($i = 0; $i < self::TEST_REPEATS; $i++) {
                        $iv = $usesIv ? random_bytes(openssl_cipher_iv_length($cipher)) : null;

                        $input = gmp_strval(gmp_random_range(
                            gmp_init(self::MIN_VALUE, 10),
                            gmp_init(self::MAX_VALUE, 10)
                        ), $base);

                        $encrypted1 = $cryptomute->encrypt($input, $base, false, 'foo', $iv);
                        $encrypted2 = $cryptomute->encrypt($input, $base, false, 'bar', $iv);
                        $this->assertNotEquals($encrypted1, $encrypted2, sprintf(
                            $message,
                            $cipher,
                            $rounds,
                            $base,
                            $input
                        ));
                    }
                }
            }
        }
    }

    /**
     * @throws \Exception
     */
    public function testDecodeWithDiffrentPasswordsProducesDiffrentResults()
    {
        $message = '[%s / %d rounds] Same base %d value decoded with diffrent passwords produce diffrent results [%s].';

        foreach (self::$testedCiphers as $cipher => $usesIv) {
            foreach (self::$testedRounds as $rounds => $description) {
                $cryptomute = $this->getCryptomute($cipher, $rounds);

                foreach (Cryptomute::$allowedBases as $base => $pattern) {
                    for ($i = 0; $i < self::TEST_REPEATS; $i++) {
                        $iv = $usesIv ? random_bytes(openssl_cipher_iv_length($cipher)) : null;

                        $input = gmp_strval(gmp_random_range(
                            gmp_init(self::MIN_VALUE, 10),
                            gmp_init(self::MAX_VALUE, 10)
                        ), $base);

                        $decrypted1 = $cryptomute->decrypt($input, $base, false, 'foo', $iv);
                        $decrypted2 = $cryptomute->decrypt($input, $base, false, 'bar', $iv);

                        $this->assertNotEquals($decrypted1, $decrypted2, sprintf(
                            $message,
                            $cipher,
                            $rounds,
                            $base,
                            $input
                        ));
                    }
                }
            }
        }
    }

    /**
     * @throws \Exception
     */
    public function testInputLessThanMinValueThrowsException()
    {
        $this->expectException(InvalidArgumentException::class);
        $cryptomute = $this->getCryptomute('aes-256-cbc', 3);
        $cryptomute->setValueRange(5000, 10000);
        $iv = random_bytes(openssl_cipher_iv_length('aes-256-cbc'));

        $cryptomute->encrypt(4999, 10, true, 'foo', $iv);
    }

    /**
     * @throws \Exception
     */
    public function testInputMoreThanMaxValueThrowsException()
    {
        $this->expectException(InvalidArgumentException::class);
        $cryptomute = $this->getCryptomute('aes-256-cbc', 3);
        $cryptomute->setValueRange(5000, 10000);
        $iv = random_bytes(openssl_cipher_iv_length('aes-256-cbc'));

        $cryptomute->encrypt(10001, 10, true, 'foo', $iv);
    }

    /**
     * @param  string  $cipher
     * @param  int  $rounds
     * @param  string|null  $minVal
     * @param  string|null  $maxVal
     *
     * @return Cryptomute
     * @throws \Exception
     */
    private function getCryptomute($cipher, $rounds, $minVal = null, $maxVal = null)
    {
        $cryptomute = new Cryptomute(
            $cipher,
            random_bytes(32),
            $rounds
        );

        return $cryptomute->setValueRange(
            $minVal ?: self::MIN_VALUE,
            $maxVal ?: self::MAX_VALUE
        );
    }
}


lib/Cryptomute/DataConverter.php
<?php

namespace Cryptomute;

/**
 * Cryptomute.
 *
 * (c) 2016 Piotr Gołębiewski
 *
 * Released under MIT license.
 * https://github.com/loostro/cryptomute
 *
 * Helper class converting data from/to various formats.
 */
class DataConverter
{
    /**
     * Converts binary string into decimal string.
     *
     * @param  string  $bin Binary string.
     * @param  int  $length Minimum output length.
     *
     * @return string Decimal string.
     */
    public static function binToDec(string $bin, int $length = 0): string
    {
        $gmp = gmp_init($bin, 2);
        $dec = gmp_strval($gmp, 10);

        return self::pad($dec, $length);
    }

    /**
     * Converts binary string into hexadecimal string.
     *
     * @param  string  $bin Binary string.
     * @param  int  $length Minimum output length.
     *
     * @return string Hexadecimal string.
     */
    public static function binToHex(string $bin, int $length = 0): string
    {
        $gmp = gmp_init($bin, 2);
        $hex = gmp_strval($gmp, 16);

        return self::pad($hex, $length);
    }

    /**
     * Converts binary string into stream of bytes.
     *
     * @param  string  $bin Binary string.
     *
     * @return mixed Stream of bytes.
     */
    public static function binToRaw(string $bin)
    {
        $gmp = gmp_init($bin, 2);

        return gmp_export($gmp);
    }

    /**
     * Converts decimal string into binary string.
     *
     * @param  string  $dec Decimal string.
     * @param  int  $length Minimum output length.
     *
     * @return string Binary string.
     */
    public static function decToBin(string $dec, int $length = 0): string
    {
        $gmp = gmp_init($dec, 10);
        $bin = gmp_strval($gmp, 2);

        return self::pad($bin, $length);
    }

    /**
     * Converts decimal string into hexadecimal string.
     *
     * @param  string  $dec Decimal string.
     * @param  int  $length Minimum output length.
     *
     * @return string Hexadecimal string.
     */
    public static function decToHex(string $dec, int $length = 0): string
    {
        $gmp = gmp_init($dec, 10);
        $hex = gmp_strval($gmp, 16);

        return self::pad($hex, $length);
    }

    /**
     * Converts decimal string into stream of bytes.
     *
     * @param  string  $dec Decimal string.
     *
     * @return mixed Stream of bytes.
     */
    public static function decToRaw(string $dec)
    {
        $gmp = gmp_init($dec, 10);

        return gmp_export($gmp);
    }

    /**
     * Converts hexadecimal string into binary string.
     *
     * @param  string  $hex Hexadecimal string.
     * @param  int  $length Minimum output length.
     *
     * @return string Binary string.
     */
    public static function hexToBin(string $hex, int $length = 0)
    {
        $gmp = gmp_init($hex, 16);
        $bin = gmp_strval($gmp, 2);

        return self::pad($bin, $length);
    }

    /**
     * Converts hexadecimal string into decimal string.
     *
     * @param  string  $hex Hexadecimal string.
     * @param  int  $length Minimum output length.
     *
     * @return string Decimal string.
     */
    public static function hexToDec(string $hex, int $length = 0): string
    {
        $gmp = gmp_init($hex, 16);
        $dec = gmp_strval($gmp, 10);

        return self::pad($dec, $length);
    }

    /**
     * Converts hexadecimal string into stream of bytes.
     *
     * @param  string  $hex Hexadecimal string.
     *
     * @return mixed Stream of bytes.
     */
    public static function hexToRaw(string $hex)
    {
        $gmp = gmp_init($hex, 16);

        return gmp_export($gmp);
    }

    /**
     * Converts stream of bytes into binary string.
     *
     * @param  mixed  $raw Stream of bytes.
     * @param  int  $length Minimum output length.
     *
     * @return string Binary string.
     */
    public static function rawToBin(string $raw, int $length = 0): string
    {
        $gmp = gmp_import($raw);
        $bin = gmp_strval($gmp, 2);

        return self::pad($bin, $length);
    }

    /**
     * Converts stream of bytes into decimal string.
     *
     * @param  mixed  $raw Stream of bytes.
     * @param  int  $length Minimum output length.
     *
     * @return string Decimal string.
     */
    public static function rawToDec($raw, int $length = 0): string
    {
        $gmp = gmp_import($raw);
        $dec = gmp_strval($gmp, 10);

        return self::pad($dec, $length);
    }

    /**
     * Converts stream of bytes into hexadecimal string.
     *
     * @param  mixed  $raw Stream of bytes.
     * @param  int  $length Minimum output length.
     *
     * @return string Hex string.
     */
    public static function rawToHex($raw, int $length = 0): string
    {
        $gmp = gmp_import($raw);
        $hex = gmp_strval($gmp, 16);

        return self::pad($hex, $length);
    }

    /**
     * Pad left with zeroes to match given length.
     *
     * @param  string  $input
     * @param  int  $length
     *
     * @return string
     */
    public static function pad(string $input, int $length = 0): string
    {
        $input = ltrim($input, '0');

        if ($input === '') {
            $input = '0';
        }

        return str_pad($input, $length, '0', STR_PAD_LEFT);
    }
}


lib/Cryptomute/Cryptomute.php
<?php

namespace Cryptomute;

use InvalidArgumentException;
use LogicException;

/**
 * Cryptomute.
 *
 * (c) 2016 Piotr Gołębiewski
 *
 * Released under MIT license.
 * https://github.com/loostro/cryptomute
 *
 * Format preserving data encryption and decryption based on feistel network.
 */
class Cryptomute
{
    public const KEY_MIN_LENGTH = 16;
    public const MIN_ROUNDS = 3;
    public const DEFAULT_MIN_VALUE = '0';
    public const DEFAULT_MAX_VALUE = '99999999999999999999'; // 20 digits
    /**
     * @var array
     */
    public static $allowedCiphers = [
        'des-cbc' => ['iv' => true, 'length' => 64],
        'aes-128-cbc' => ['iv' => true, 'length' => 128],
        'aes-192-cbc' => ['iv' => true, 'length' => 192],
        'aes-256-cbc' => ['iv' => true, 'length' => 256],
        'camellia-128-cbc' => ['iv' => true, 'length' => 128],
        'camellia-192-cbc' => ['iv' => true, 'length' => 192],
        'camellia-256-cbc' => ['iv' => true, 'length' => 256],
    ];
    /**
     * @var array
     */
    public static $allowedBases = [
        2 => '/^[0-1]+$/',
        10 => '/^[0-9]+$/',
        16 => '/^[a-f0-9]+$/',
    ];
    /**
     * @var string
     */
    protected $minValue;
    /**
     * @var string
     */
    protected $maxValue;
    /**
     * @var string
     */
    protected $cipher;
    /**
     * @var string
     */
    protected $password;
    /**
     * @var string
     */
    protected $key;
    /**
     * @var array
     */
    protected $roundKeys;
    /**
     * @var string
     */
    protected $iv;
    /**
     * @var int
     */
    protected $rounds;
    /**
     * @var int
     */
    protected $cipherLength;
    /**
     * @var int
     */
    protected $blockSize;
    /**
     * @var int
     */
    protected $binSize;
    /**
     * @var int
     */
    protected $decSize;
    /**
     * @var int
     */
    protected $hexSize;
    /**
     * @var int
     */
    protected $sideSize;

    /**
     * Cryptomute constructor.
     *
     * @param  string  $cipher Cipher used to encrypt.
     * @param  string  $baseKey Base key, from which all round keys are derrived.
     * @param  int  $rounds Number of rounds.
     *
     * @throws InvalidArgumentException If provided invalid constructor parameters.
     * @throws LogicException           If side size is longer than cipher length.
     */
    public function __construct(string $cipher, string $baseKey, int $rounds = 3)
    {
        if (! array_key_exists($cipher, self::$allowedCiphers)) {
            throw new InvalidArgumentException(sprintf(
                'Cipher must be one of "%s".',
                implode(', ', array_keys(self::$allowedCiphers))
            ));
        }

        $this->cipher = $cipher;
        $this->cipherLength = self::$allowedCiphers[$cipher]['length'];
        $this->setValueRange(self::DEFAULT_MIN_VALUE, self::DEFAULT_MAX_VALUE);

        if (! is_int($rounds) || $rounds < self::MIN_ROUNDS || $rounds % 2 !== 1) {
            throw new InvalidArgumentException(sprintf(
                'Number of rounds must be an odd integer greater or equal %d',
                self::MIN_ROUNDS
            ));
        }

        $this->rounds = $rounds;

        if (strlen($baseKey) < self::KEY_MIN_LENGTH) {
            throw new InvalidArgumentException(sprintf(
                'Key must be at least %d characters long.',
                self::KEY_MIN_LENGTH
            ));
        }

        $this->key = $baseKey;
    }

    /**
     * Set value range and pad sizes.
     *
     * @param  string  $minValue Minimum value. String representation of positive integer value or zero.
     * @param  string  $maxValue Maximum value. String representation of positive integer value.
     *
     * @return Cryptomute
     * @throws InvalidArgumentException If provided invalid parameters.
     */
    public function setValueRange(string $minValue, string $maxValue): self
    {
        if (preg_match('/^([1-9][0-9]*)|([0]{1})$/', $minValue) !== 1) {
            throw new InvalidArgumentException(
                'Min value must contain only digits.'
            );
        }

        if (preg_match('/^[1-9][0-9]*$/', $maxValue) !== 1) {
            throw new InvalidArgumentException(
                'Max value must start with a nonzero digit and contain only digits.'
            );
        }

        if (gmp_cmp($maxValue, $minValue) < 1) {
            throw new InvalidArgumentException('Max value must be greater than min value.');
        }

        $this->minValue = $minValue;
        $this->maxValue = $maxValue;

        // find the minimum number of even bits to span whole set
        $this->binSize = 2;
        $span = gmp_init('4', 10);
        $multiplier = gmp_init('4', 10);
        do {
            $this->binSize += 2;
            $span = gmp_mul($span, $multiplier);
        } while (gmp_cmp($span, $this->maxValue) < 1);

        $this->decSize = strlen($this->maxValue);
        $this->hexSize = $this->binSize / 4;

        $this->sideSize = (int) $this->binSize / 2;

        if ($this->sideSize > $this->cipherLength) {
            throw new LogicException(sprintf(
                'Side size (%d bits) must be less or equal to cipher length (%d bits)',
                $this->sideSize,
                $this->cipherLength
            ));
        }

        return $this;
    }

    /**
     * Encrypts input data. Acts as a public alias for _encryptInternal method.
     *
     * @param  string  $input String representation of input number.
     * @param  int  $base Input number base.
     * @param  bool  $pad Pad left with zeroes?
     * @param  string|null  $password Encryption password.
     * @param  string|null  $iv Encryption initialization vector. Must be unique!
     *
     * @return string Outputs encrypted data in the same format as input data.
     */
    public function encrypt(string $input, int $base = 10, bool $pad = false, $password = null, $iv = null): string
    {
        return $this->_encryptInternal($input, $base, $pad, $password, $iv, true);
    }

    /**
     * Encrypts input data.
     *
     * @param  string  $input String representation of input number.
     * @param  int  $base Input number base.
     * @param  bool  $pad Pad left with zeroes?
     * @param  string|null  $password Encryption password.
     * @param  string|null  $iv Encryption initialization vector. Must be unique!
     * @param  bool  $checkVal Should check if input value is in range?
     *
     * @return string Outputs encrypted data in the same format as input data.
     */
    private function _encryptInternal(string $input, int $base, bool $pad, $password, $iv, $checkVal = false): string
    {
        $this->_validateInput($input, $base, $checkVal);
        $this->_validateIv($iv);
        $hashPassword = $this->_hashPassword($password);
        $roundKeys = $this->_roundKeys($hashPassword, $iv);

        $binary = $this->_convertToBin($input, $base);

        for ($i = 1; $i <= $this->rounds; $i++) {
            $left = substr($binary, 0, $this->sideSize);
            $right = substr($binary, -1 * $this->sideSize);

            $key = $roundKeys[$i];
            $round = $this->_round($right, $key, $hashPassword, $iv);

            $newLeft = $right;
            $newRight = $this->_binaryXor($left, $round);

            $binary = $newLeft.$newRight;
        }

        $output = $this->_convertFromBin($binary, $base, $pad);
        $compare = DataConverter::binToDec($binary);

        return (gmp_cmp($this->minValue, $compare) > 0 || gmp_cmp($compare, $this->maxValue) > 0)
            ? $this->_encryptInternal($output, $base, $pad, $password, $iv, false)
            : $output;
    }

    /**
     * Decrypts input data.
     *
     * @param  string  $input Encrypted input.
     * @param  int  $base Input data base.
     * @param  bool  $pad Pad left with zeroes?
     * @param  string|null  $password Decryption password.
     * @param  string|null  $iv Decryption initialization vector.
     *
     * @return string Outputs encrypted data in the same format as input data.
     */
    public function decrypt(string $input, int $base = 10, bool $pad = false, $password = null, $iv = null): string
    {
        $this->_validateInput($input, $base);
        $this->_validateIv($iv);
        $hashPassword = $this->_hashPassword($password);
        $roundKeys = $this->_roundKeys($hashPassword, $iv);

        $binary = $this->_convertToBin($input, $base);

        for ($i = $this->rounds; $i > 0; $i--) {
            $left = substr($binary, 0, $this->sideSize);
            $right = substr($binary, -1 * $this->sideSize);

            $key = $roundKeys[$i];
            $round = $this->_round($left, $key, $hashPassword, $iv);

            $newLeft = $this->_binaryXor($right, $round);
            $newRight = $left;

            $binary = $newLeft.$newRight;
        }

        $output = $this->_convertFromBin($binary, $base, $pad);
        $compare = DataConverter::binToDec($binary);

        return (gmp_cmp($this->minValue, $compare) > 0 || gmp_cmp($compare, $this->maxValue) > 0)
            ? $this->decrypt($output, $base, $pad, $password, $iv)
            : $output;
    }

    /**
     * Encrypt helper.
     *
     * @param  string  $input
     * @param  string  $password
     * @param  string|null  $iv
     *
     * @return string Steam of encrypted bytes.
     */
    private function _encrypt(string $input, string $password, $iv = null): string
    {
        return openssl_encrypt($input, $this->cipher, $password, true, $iv);
    }

    /**
     * Round function helper.
     *
     * @param  string  $input
     * @param  string  $key
     * @param  string  $hashPassword
     * @param  string|null  $iv
     *
     * @return string Binary string.
     */
    private function _round(string $input, string $key, string $hashPassword, $iv = null): string
    {
        $bin = DataConverter::rawToBin($this->_encrypt($input.$key, $hashPassword, $iv));

        return substr($bin, -1 * $this->sideSize);
    }

    /**
     * Binary xor helper.
     *
     * @param  string  $left
     * @param  string  $round
     *
     * @return string Binary string.
     */
    private function _binaryXor(string $left, string $round): string
    {
        $xOr = gmp_xor(
            gmp_init($left, 2),
            gmp_init($round, 2)
        );

        $bin = gmp_strval($xOr, 2);

        return str_pad($bin, $this->sideSize, '0', STR_PAD_LEFT);
    }

    /**
     * Helper method converting input data to binary string.
     *
     * @param  string  $input
     * @param  string  $base
     *
     * @return string
     */
    private function _convertToBin(string $input, string $base): string
    {
        switch ($base) {
            case 2:
                return DataConverter::pad($input, $this->binSize);
            case 10:
                return DataConverter::decToBin($input, $this->binSize);
            case 16:
                return DataConverter::hexToBin($input, $this->binSize);
        }
    }

    /**
     * Helper method converting input data from binary string.
     *
     * @param  string  $binary
     * @param  string  $base
     * @param  string  $pad
     *
     * @return string
     */
    private function _convertFromBin(string $binary, string $base, string $pad): string
    {
        switch ($base) {
            case 2:
                return DataConverter::pad($binary, ($pad ? $this->binSize : 0));
            case 10:
                return DataConverter::binToDec($binary, ($pad ? $this->decSize : 0));
            case 16:
                return DataConverter::binToHex($binary, ($pad ? $this->hexSize : 0));
        }
    }

    /**
     * Validates input data.
     *
     * @param  string  $input
     * @param  string  $base
     * @param  bool  $checkDomain Should check if input is in domain?
     *
     * @throws InvalidArgumentException If provided invalid type.
     */
    private function _validateInput(string $input, string $base, bool $checkDomain = false): void
    {
        if (! array_key_exists($base, self::$allowedBases)) {
            throw new InvalidArgumentException(sprintf(
                'Type must be one of "%s".',
                implode(', ', array_keys(self::$allowedBases))
            ));
        }

        if (preg_match(self::$allowedBases[$base], $input) !== 1) {
            throw new InvalidArgumentException(sprintf(
                'Input data "%s" does not match pattern "%s".',
                $input,
                self::$allowedBases[$base]
            ));
        }

        if ($checkDomain) {
            $compare = gmp_init($input, $base);

            if (gmp_cmp($this->minValue, $compare) > 0 || gmp_cmp($compare, $this->maxValue) > 0) {
                throw new InvalidArgumentException(sprintf(
                    'Input value "%d" is out of domain range "%d - %d".',
                    gmp_strval($compare, 10),
                    $this->minValue,
                    $this->maxValue
                ));
            }
        }
    }

    /**
     * Validates initialization vector.
     *
     * @param  string|null  $iv
     */
    private function _validateIv($iv = null): void
    {
        if (self::$allowedCiphers[$this->cipher]['iv']) {
            $this->blockSize = openssl_cipher_iv_length($this->cipher);

            $ivLength = mb_strlen($iv, '8bit');
            if ($ivLength !== $this->blockSize) {
                throw new InvalidArgumentException(sprintf(
                    'Initialization vector of %d bytes is required for cipher "%s", %d given.',
                    $this->blockSize,
                    $this->cipher,
                    $ivLength
                ));
            }
        }
    }

    /**
     * Hashes the password.
     *
     * @param  string|null  $password
     *
     * @return string
     */
    private function _hashPassword($password = null): string
    {
        if (null !== $password) {
            $this->password = hash('sha3-512', $password);
        }

        return $this->password;
    }

    /**
     * Generates hash keys.
     *
     * @param  string|null  $hashPassword
     * @param  string|null  $iv
     *
     * @return array
     */
    private function _roundKeys($hashPassword = null, $iv = null): array
    {
        $roundKeys = [];
        $prevKey = $this->_encrypt($this->key, $hashPassword, $iv);
        for ($i = 1; $i <= $this->rounds; $i++) {
            $prevKey = $this->_encrypt($prevKey, $hashPassword, $iv);
            $roundKeys[$i] = substr(DataConverter::rawToBin($prevKey), -1 * $this->sideSize);
        }

        return $roundKeys;
    }
}
